dlab_core/infrastructure/services.py

```diff
@@ -0,0 +1,158 @@
+# *****************************************************************************
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+# ******************************************************************************
+import requests
+import urllib3
+
+from dlab_core.domain.exceptions import DLabException
+
+from jose import jwt
+
+
+URL_INTROSPECT = '{realm_address}/protocol/openid-connect/token/introspect'
+
+
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+
+
+class KeyCloakError(DLabException):
+    pass
+
+
+class InvalidParameterFormatError(KeyCloakError):
+    message = 'Parameter {0} is not of {1} type: {2}'
+
+    def __str__(self):
+        return self.message.format(*self.args)
+
+
+class KeyCloakConnectionError(KeyCloakError):
+    pass
+
+
+class KeyCloak(object):
+    """Client for KeyCloak interaction
+
+    :type keycloak_host: str
+    :param keycloak_host: IP or hostname for KeyCloak server to use
+
+    :type realm_name: str
+    :param realm_name: name of the KeyCloak Realm to use
+
+    :type client_id: str
+    :param client_id: KeyCloak Client to use for token introspection
+
+    :type client_secret: str
+    :param client_secret: KeyCloak Client secret string
+    """
+    def __init__(self, keycloak_host, realm_name, client_id, client_secret):
+        check = ['keycloak_host', 'realm_name', 'client_id', 'client_secret']
+        for arg_name in check:
+            arg_value = locals()[arg_name]
+            if not isinstance(arg_value, str):
+                raise InvalidParameterFormatError(
+                    arg_name, 'str', type(arg_value).__name__
+                )
+        self._realm_address = '{}/realms/{}'.format(
+            keycloak_host.rstrip('/'), realm_name
+        )
+        self._client_id = client_id
+        self._client_secret = client_secret
+        self._pub_key = self.get_key()
```

Comment:
there is special construction for this
https://stackoverflow.com/questions/4558763/how-to-lazy-load-a-data-structure-python

Reply:
Link shows __getitem__ method using. Should I use this approach or maybe property getter/setter is better?